lib/portlayer/storage/vsphere/volume.go

```diff
@@ -0,0 +1,212 @@
+
+	// wraps our vmdks and filesystem primitives.
+	dm *disk.Manager
+
+	sess *session.Session
+}
+
+func NewVolumeStore(ctx context.Context, s *session.Session) (*VolumeStore, error) {
+	dm, err := disk.NewDiskManager(ctx, s)
+	if err != nil {
+		return nil, err
+	}
+
+	v := &VolumeStore{
+		dm:   dm,
+		sess: s,
+		ds:   make(map[url.URL]*Datastore),
+	}
+
+	return v, nil
+}
+
+// AddStore adds a volumestore by uri.
+//
+// ds is the Datastore (+ path) volumes will be created under.
+//       The resulting path will be parentDir/VIC/volumes.
+// storeName is the name used to refer to the datastore + path (ds above).
+//
+// returns the URL used to refer to the volume store
+func (v *VolumeStore) AddStore(ctx context.Context, ds *Datastore, storeName string) (*url.URL, error) {
+	u, err := util.VolumeStoreNameToURL(storeName)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := v.ds[*u]; ok {
+		return nil, fmt.Errorf("volumestore (%s) already added", u.String())
+	}
+
+	if _, err = ds.Mkdir(ctx, true, volumesDir); err != nil {
+		return nil, err
+	}
+
+	v.ds[*u] = ds
+	return u, nil
+}
+
+func (v *VolumeStore) getDatastore(store *url.URL) (*Datastore, error) {
+	// find the datastore
+	dstore, ok := v.ds[*store]
+	if !ok {
+		return nil, fmt.Errorf("volumestore (%s) not found", store.String())
+	}
+
+	return dstore, nil
+}
+
+// Returns the path to the vol relative to the given store.  The dir structure
+// for a vol in the datastore is `<configured datastore path>/volumes/<vol ID>/<vol ID>.vmkd`.
+// Everything up to "volumes" is taken care of by the datastore wrapper.
+func (v *VolumeStore) volDirPath(ID string) string {
+	return path.Join(volumesDir, ID)
+}
+
+// Returns the path to the vmdk itself (in datastore URL format)
+func (v *VolumeStore) volDiskDsURL(store *url.URL, ID string) (string, error) {
+	// find the datastore
+	dstore, err := v.getDatastore(store)
+	if err != nil {
+		return "", err
+	}
+
+	// XXX this could be hidden in a helper.  We shouldn't use rooturl outside the datastore struct
+	return path.Join(dstore.RootURL, v.volDirPath(ID), ID+".vmdk"), nil
+}
+
+func (v *VolumeStore) VolumeCreate(ctx context.Context, ID string, store *url.URL, capacityMB uint64, info map[string][]byte) (*storage.Volume, error) {
+
+	// find the datastore
+	dstore, err := v.getDatastore(store)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the volume directory in the store.
+	_, err = dstore.Mkdir(ctx, false, v.volDirPath(ID))
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the path to the disk in datastore uri format
+	var volDiskDsURL string
+	volDiskDsURL, err = v.volDiskDsURL(store, ID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the disk
+	vmdisk, err := v.dm.CreateAndAttach(ctx, volDiskDsURL, "", int64(capacityMB), os.O_RDWR)
+	if err != nil {
+		return nil, err
+	}
+	defer v.dm.Detach(ctx, vmdisk)
+
+	// Make the filesystem and set its label to the volume ID
+	if err = vmdisk.Mkfs(ID); err != nil {
```

Comment:
Shouldn't this be set to Label instead so we can identify the disk in the containerVMs?